source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst

```diff
@@ -0,0 +1,156 @@
+
+   public class MyPoint2D extends ComplexData<MyPoint2D> {
+
+      private final double x;
+      private final double y;
+
+      //Constructor should take all the different fields needed and assign them their corresponding instance variables.
+      public MyPoint2D(double x, double y) {
+         this.x = x;
+         this.y = y;
+      }
+
+      @Override
+      public Map<String, Object> asMap() {
+         return null;
+      }
+   }
+
+It is also good practise to override the default ``equals`` and ``hashcode`` methods to ensure that different objects are considered equivalent when their fields are the same.
+The ``asMap()`` method should return the data represented in a simple Map object as it will be mapped to the NetworkTable entry it corresponds to. In this case, we can represent the point as its X and Y coordinates and return a ``Map`` containing them.
+
+.. code-block:: java
+
+   import edu.wpi.first.shuffleboard.api.data.ComplexData;
+
+   import java.util.Map;
+
+   public class MyPoint2D extends ComplexData<MyPoint2D> {
+
+      private final double x;
+      private final double y;
+
+      //Constructor should take all the different fields needed and assign them to their corresponding instance variables.
+      public Point(double x, double y) {
+         this.x = x;
+         this.y = y;
+      }
+
+      @Override
+      public Map<String, Object> asMap() {
+         return Map.of("x", this.x, "y", this.y);
+      }
+
+      @Override
+      public boolean equals(MyPoint2D point) {
+         // Checks this object and point object for equal x and y values
+         if (this.x == point.asMap().get("x") && this.y == point.asMap().get("y")) {
+            return true;
+         } else {
+            return false;
+         }
+      }
+    }
+
+Other methods can be added to retrieve or edit fields and instance variables, however, it is good practise to make these classes immutable to prevent changing the source data objects.
+Instead, you can make a new copy object instead of manipulating the existing object. For example, if we wanted to change the y coordinate of our point, we can define the following method:
+
+.. code-block:: java
+
+   public MyPoint2D withX(double newY) {
+      return new MyPoint2D(this.x, newY);
+   }
+
+This creates a new ``MyPoint2D`` object and returns it with the new y-coordinate.
+
+Creating a Data Types
+---------------------
+There are two different data types that can be made, Simple data types involve data that contains only one field (ie. a single number or string) whereas Complex data types require multiple data fields (ie. multiple strings, multiple numbers).
+
+In order to define a simple data type, the class must extend the ``SimpleDataType<DataType>`` class with the data type needed and implement the ``getDefaultValue()`` method. In this example, we will use a double as our simple data type.
+
+.. code-block:: java
+
+   public final class MyDataType extends SimpleDataType<double> {
```

Comment:
Naming should be `DoubleType` (or maybe `ExampleDoubleType`). `MyDataType` is a poor name when it's for a concrete example.

Can't do generics on primitive types; you need to use the boxed class `Double`